src/Interpreters/Cache/QueryResultCache.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <Core/Block.h>
+#include <Core/Settings.h>
+#include <Parsers/IAST.h>
+#include <Processors/Chunk.h>
+#include <QueryPipeline/Pipe.h>
+
+namespace DB
+{
+
+/// Does AST contain non-deterministic functions like rand() and now()?
+bool astContainsNonDeterministicFunctions(ASTPtr ast, ContextPtr context);
+
+/// Maps queries to query results. Useful to avoid repeated query calculation.
+///
+/// The cache does not aim to be transactionally consistent (which is difficult to get right). For example, the cache is not invalidated
+/// when data is inserted/deleted into/from tables referenced by queries in the cache. In such situations, incorrect results may be
+/// returned. In order to still obtain sufficiently up-to-date query results, a expiry time must be specified for each cache entry after
+/// which it becomes stale and is ignored. Stale entries are removed opportunistically from the cache, they are only evicted when a new
+/// entry is inserted and the cache has insufficient capacity.
+class QueryResultCache
+{
+public:
+    /// Represents a query result in the cache.
+    struct Key
+    {
+        /// ----------------------------------------------------
+        /// The actual key (data which gets hashed):
+
+        /// Unlike the query string, the AST is agnostic to lower/upper case (SELECT vs. select)
+        const ASTPtr ast;
+
+        /// Identifies a (virtual) cache partition. Can be used to cache the same query multiple times with different timeouts.
+        const String partition_key;
+
+        /// Note: For a transactionally consistent cache, we would need to include the system settings in the cache key or invalidate the
+        /// cache whenever the settings change. This is because certain settings (e.g. "additional_table_filters") can affect the query
+        /// result.
+
+        /// ----------------------------------------------------
+        /// Additional stuff data stored in the key, not hashed:
+
+        /// For constructing the pipe.
+        const Block header;
+
+        /// Std::nullopt means that the associated entry can be read by other users. In general, sharing is a bad idea: First, it is
+        /// unlikely that different users pose the same queries. Second, sharing potentially breaches security. E.g. User A should not be
+        /// able to bypass row policies on some table by running the same queries as user B for whom no row policies exist.
+        const std::optional<String> username;
+
+        /// When does the entry expire?
+        const std::chrono::time_point<std::chrono::system_clock> expires_at;
```

Comment:
I generally like this idea, it is a bit more intuitive. For each cache entry, we would store the insertion time instead of the expiry time. Cache lookups can then check if a found entry is recent enough wrt. a user-supplied freshness.

The downside is that eviction due to space shortage becomes more tricky. Right now, we know if a cache entry is definitely stale (based on the expiry time) and can evict it. So this is a very simple strategy. But if insertion timestamps are stored, we cannot know if there will be future queries which consider the entry fresh enough. Heuristics are needed: evict the largest entries, the least recently used entries, the oldest entries (or at least the entries older than the default-configured freshness), etc.

I need to think a bit more about it and would like to put that into my backlog (--> do it separately).